### PR TITLE
Liquid plugin

### PR DESCRIPTION
**Pull request by alanz**
Add a plugin for liquid haskell.

This is disabled by default, and has to be enabled via a config setting (similar to `hlintOn`)

The `diagnostic provider` interface has changed, to include the concept of an async provider, which can run an external process, and then send results when it finishes, without blocking the dispatcher.

### CODE REVIEW

File: src/Haskell/Ide/Engine/LSP/Config.hs [hunk 1]
```diff
@@ -34,3 +42,20 @@ instance FromJSON Config where
...
 --                 {_settings = Object (fromList [("languageServerHaskell",Object (fromList [("hlintOn",Bool True)
 --                                                                                          ,("maxNumberOfProblems",Number 100.0)]))])}}
+
+instance ToJSON Config where
+  toJSON (Config h m l) = object [ "languageServerHaskell" .= r ]
+    where
+      r = object [ "hlintOn"             .= h
+                 , "maxNumberOfProblems" .= m
+                 , "liquidOn"            .= l
+                 ]
```
Comment (lorenzo) [inline src/Haskell/Ide/Engine/LSP/Config.hs]: Can we add a unit test showing that `FromJSON` and `ToJSON` are symmetrical? I often run into bugs when I only change one of the instances and not the other

File: src/Haskell/Ide/Engine/LSP/Config.hs [hunk 2]
```diff
@@ -34,3 +42,20 @@ instance FromJSON Config where
...
+
+-- ---------------------------------------------------------------------
+
+-- | For the diagnostic providers in the config, return a map of
+-- current enabled state, indexed by the plugin id.
+getDiagnosticProvidersConfig :: Config -> Map.Map PluginId Bool
+getDiagnosticProvidersConfig c = Map.fromList [("applyrefact",hlintOn c)
+                                              ,("liquid",     liquidOn c)
```
Comment (lorenzo) [inline src/Haskell/Ide/Engine/LSP/Config.hs]: Is this brittany's formatting? looks off

File: src/Haskell/Ide/Engine/Plugin/Liquid.hs [hunk 3]
```diff
@@ -4,8 +4,10 @@
 {-# LANGUAGE NamedFieldPuns      #-}
 module Haskell.Ide.Engine.Plugin.Liquid where
 
+import           Control.Concurrent.Async
 import           Control.Monad
 import           Control.Monad.IO.Class
+import           Control.Exception (bracket)
 #if __GLASGOW_HASKELL__ < 804
```
Comment (lorenzo) [inline src/Haskell/Ide/Engine/Plugin/Liquid.hs]: This line can now be removed

File: src/Haskell/Ide/Engine/Plugin/Liquid.hs [hunk 4]
```diff
@@ -32,11 +38,13 @@ liquidDescriptor plId = PluginDescriptor
...
   , pluginDesc = "Integration with Liquid Haskell"
   , pluginCommands =
-      [ PluginCommand "sayHello" "say hello" sayHelloCmd
-      , PluginCommand "sayHelloTo ""say hello to the passed in param" sayHelloToCmd
+      [ PluginCommand "sayHello"   "say hello"                        sayHelloCmd
+      , PluginCommand "sayHelloTo" "say hello to the passed in param" sayHelloToCmd
       ]
   , pluginCodeActionProvider = Nothing
-  , pluginDiagnosticProvider = Just (DiagnosticProvider (S.singleton DiagnosticOnSave) diagnosticProvider)
+  , pluginDiagnosticProvider= Just (DiagnosticProvider
```
Comment (lorenzo) [inline src/Haskell/Ide/Engine/Plugin/Liquid.hs]: Alignment looks off here as well

File: test/functional/LiquidSpec.hs [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+-- import           Language.Haskell.LSP       as LSP
+import           Language.Haskell.LSP.Types as LSP hiding (contents, error )
+import           Haskell.Ide.Engine.LSP.Config
+import           Test.Hspec
+import           TestUtils
+import           Utils
+
+-- ---------------------------------------------------------------------
+
+spec :: Spec
```
Comment (lorenzo) [inline test/functional/LiquidSpec.hs]: Can this be guarded by LH being installed? That or we need to also install it in travis

File: src/Haskell/Ide/Engine/LSP/Config.hs (hunk 1 above)
Comment (lorenzo) [reply src/Haskell/Ide/Engine/LSP/Config.hs]: Nevermind, I read further and the test seems to be present already

File: src/Haskell/Ide/Engine/Plugin/Liquid.hs (hunk 3 above)
Comment (alanz) [reply src/Haskell/Ide/Engine/Plugin/Liquid.hs]: ok

Comment (alanz) [reply src/Haskell/Ide/Engine/Plugin/Liquid.hs]: Except it fails -Wall on GHC 8.4.3 if the #if is removed

File: src/Haskell/Ide/Engine/LSP/Config.hs (hunk 1 above)
Comment (Bubba) [reply src/Haskell/Ide/Engine/LSP/Config.hs]: Why does this need a `ToJSON` instance? Not opposed to it just curious

Comment (alanz) [reply src/Haskell/Ide/Engine/LSP/Config.hs]: For the quickcheck roundtrip tests

**Comment by alanz** [comment]
@Bubba updating the default logging to be off is fine.

In practical terms, I will still have two configs, one for logging, one for not, so that I can turn it on while debugging an individual test.

**Pull request merged**